npm/qsharp/ux/circuit-vis/sqore.ts

```diff
@@ -336,41 +449,102 @@ export class Sqore {
-    operations.forEach((op) => {
-      if (op.conditionalRender === ConditionalRender.AsGroup)
-        this.expandOperation(op.children || [], id);
-      if (op.dataAttributes == null) return op;
-      const opId: string = op.dataAttributes["id"];
-      if (opId === id && op.children != null) {
-        op.conditionalRender = ConditionalRender.AsGroup;
-        op.dataAttributes["expanded"] = "true";
-      }
-    });
+  private expandOperation(
+    componentGrid: ComponentGrid,
+    location: string,
+  ): void {
+    componentGrid.forEach((col) =>
+      col.components.forEach((op) => {
+        if (op.conditionalRender === ConditionalRender.AsGroup)
+          this.expandOperation(op.children || [], location);
+        if (op.dataAttributes == null) return op;
+        const opId: string = op.dataAttributes["location"];
+        if (opId === location && op.children != null) {
+          op.conditionalRender = ConditionalRender.AsGroup;
+          op.dataAttributes["expanded"] = "true";
+        }
+      }),
+    );
   }
 
   /**
    * Collapse selected operation for zoom-out interaction.
    *
-   * @param operations List of circuit operations.
-   * @param id ID of operation to collapse.
+   * @param componentGrid Grid of circuit components.
+   * @param parentLoc Location of operation to collapse.
    *
    */
-  private collapseOperation(operations: Operation[], parentId: string): void {
-    operations.forEach((op) => {
-      if (op.conditionalRender === ConditionalRender.AsGroup)
-        this.collapseOperation(op.children || [], parentId);
-      if (op.dataAttributes == null) return op;
-      const opId: string = op.dataAttributes["id"];
-      // Collapse parent gate and its children
-      if (opId.startsWith(parentId)) {
-        op.conditionalRender = ConditionalRender.Always;
-        delete op.dataAttributes["expanded"];
-      }
+  private collapseOperation(
+    componentGrid: ComponentGrid,
+    parentLoc: string,
+  ): void {
+    componentGrid.forEach((col) =>
+      col.components.forEach((op) => {
+        if (op.conditionalRender === ConditionalRender.AsGroup)
+          this.collapseOperation(op.children || [], parentLoc);
+        if (op.dataAttributes == null) return op;
+        const opId: string = op.dataAttributes["location"];
+        // Collapse parent gate and its children
+        if (opId.startsWith(parentLoc)) {
+          op.conditionalRender = ConditionalRender.Always;
+          delete op.dataAttributes["expanded"];
+        }
+      }),
+    );
+  }
+
+  public useDraggable(): Sqore {
+    this.extensions = [...this.extensions, extensionDraggable];
+    return this;
+  }
+
+  public usePanel(options?: PanelOptions): Sqore {
+    this.extensions = [...this.extensions, extensionPanel(options)];
+    return this;
+  }
+
+  public useEvents(): Sqore {
+    this.extensions = [...this.extensions, extensionEvents];
+    return this;
+  }
```

Comment:
There are several things I'm not really following with this design. 

- Why do these attributes all need to be exposed as separate methods? Isn't a circuit just editable, or not editable? 

- What's the deal with `extensions`?

- I know fluent interfaces have their fans, but I'm just general confused by them. Why go with a fluent interface here? These could return void... or they could just be flags passed into the constructor.

Reply:
I've removed the extensions mechanism for extending the Sqore and just call all the components deliberately in the render function.